src/plugins/wizard/public/application/utils/get_top_nav_config.tsx

```diff
@@ -37,64 +37,73 @@ import {
 interface TopNavConfigParams {
-  visInstance: Record<string, any>; // TODO: fix this type
+  visualizationIdFromUrl: string;
+  savedWizardVis: WizardVisSavedObject;
+  visualizationState: VisualizationState;
+  styleState: StyleState;
   hasUnappliedChanges: boolean;
 }
 
-export const getTopNavconfig = (
-  { visInstance, hasUnappliedChanges }: TopNavConfigParams,
+export const getTopNavConfig = (
   {
-    savedObjects: { client: savedObjectsClient },
-    toastNotifications,
-    i18n: { Context: I18nContext },
-  }: WizardServices
+    visualizationIdFromUrl,
+    savedWizardVis,
+    visualizationState,
+    styleState,
+    hasUnappliedChanges,
+  }: TopNavConfigParams,
+  { history, toastNotifications, i18n: { Context: I18nContext } }: WizardServices
 ) => {
-  const { state } = visInstance;
   const topNavConfig: TopNavMenuData[] = [
     {
       id: 'save',
       iconType: 'save',
-      emphasize: true, // TODO: need to be conditional for save vs create (save as)?
-      description: 'Save Visualization', // TODO: i18n
+      emphasize: savedWizardVis && !savedWizardVis.id,
+      description: i18n.translate('wizard.topNavMenu.saveVisualizationButtonAriaLabel', {
+        defaultMessage: 'Save Visualization',
+      }),
       className: 'saveButton',
-      label: 'save', // TODO: i18n
+      label: i18n.translate('wizard.topNavMenu.saveVisualizationButtonLabel', {
+        defaultMessage: 'save',
+      }),
       testId: 'wizardSaveButton',
       disableButton: hasUnappliedChanges,
       tooltip() {
         if (hasUnappliedChanges) {
-          return i18n.translate('visualize.topNavMenu.saveVisualizationDisabledButtonTooltip', {
+          return i18n.translate('wizard.topNavMenu.saveVisualizationDisabledButtonTooltip', {
             defaultMessage: 'Apply aggregation configuration changes before saving', // TODO: Update text to match agg save flow
           });
         }
       },
-      run: (anchorElement) => {
+      run: (_anchorElement) => {
         const onSave = async ({
-          // TODO: Figure out what the other props here do
           newTitle,
           newCopyOnSave,
           isTitleDuplicateConfirmed,
           onTitleDuplicate,
           newDescription,
           returnToOrigin,
         }: OnSaveProps & { returnToOrigin: boolean }) => {
-          // TODO: Save the actual state of the wizard
-          const wizardSavedObject = visInstance.id
-            ? await savedObjectsClient.update(WIZARD_SAVED_OBJECT, visInstance.id, {
-                title: newTitle,
-                description: newDescription,
-                state,
-              })
-            : await savedObjectsClient.create(WIZARD_SAVED_OBJECT, {
-                title: newTitle,
-                description: newDescription,
-                state,
-              });
+          if (!savedWizardVis) {
+            return;
+          }
+          savedWizardVis.visualizationState = JSON.stringify(visualizationState);
```

Comment:
Not for this PR, but to handle a failed save and to reset the saved object, you can just use immers's `produce` function to clone the object and modify the before saving it. That way the original saved object is untouched. That also keeps `savedWizardVis` from being mutated